p-token/src/processor/set_authority.rs

```diff
@@ -108,34 +106,39 @@ pub fn process_set_authority(accounts: &[AccountInfo], instruction_data: &[u8])
     Ok(())
 }
 
-struct SetAuthority<'a> {
-    authority_type: AuthorityType,
-
-    new_authority: Option<&'a Pubkey>,
+#[repr(C)]
+struct SetAuthority {
+    authority_type: u8,
 
-    _data: PhantomData<&'a [u8]>,
+    new_authority: (u8, Pubkey),
 }
 
-impl SetAuthority<'_> {
-    #[inline(always)]
-    pub fn try_from_bytes(bytes: &[u8]) -> Result<SetAuthority, ProgramError> {
+impl SetAuthority {
+    #[inline]
+    pub fn try_from_bytes(bytes: &[u8]) -> Result<&SetAuthority, ProgramError> {
         // The minimum expected size of the instruction data is either 2 or 34 bytes:
         //   - authority_type (1 byte)
         //   - option + new_authority (1 byte + 32 bytes)
         unsafe {
             match bytes.len() {
-                2 if *bytes.get_unchecked(1) == 0 => Ok(SetAuthority {
-                    authority_type: (*bytes.get_unchecked(0)).try_into()?,
-                    new_authority: None,
-                    _data: PhantomData,
-                }),
-                34 if *bytes.get_unchecked(1) == 1 => Ok(SetAuthority {
-                    authority_type: (*bytes.get_unchecked(0)).try_into()?,
-                    new_authority: Some(&*(bytes.as_ptr().add(2) as *const Pubkey)),
-                    _data: PhantomData,
-                }),
+                2 if *bytes.get_unchecked(1) == 0 => Ok(&*(bytes.as_ptr() as *const SetAuthority)),
+                34 => Ok(&*(bytes.as_ptr() as *const SetAuthority)),
```

Comment:
Did you mean to remove the `if *bytes.get_unchecked(1) == 1` portion to this?

We probably want to keep that to be safe

Reply:
The issue is that it disproportionally increases CUs to have it. We could do the check on the accessor method:
```rust
pub fn new_authority(&self) -> Option<&Pubkey> {
    if self.new_authority.0 == 1 {
        Option::Some(&self.new_authority.1)
    } else {
        Option::None
    }
}
```

Reply:
Or even return a `Result`?

Reply:
That's fine, but it has to be `1` or `0` or it errors, which means this would need to return `Result<Option<&Pubkey>, ProgramError>` instead

Reply:
Yeah, I think that is the best option. I will test it, but I can't imagine that it would add too much overhead.

Reply:
We can't do this and the current implementation has a different behaviour than the original one. We need to validate the instruction data before executing any code of the processor, otherwise we can get different results.

I fixed both `SetAuthority` and `InitializeMint` instruction data validation, adding the validation for the option byte.